plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/SortByInputModelProcessor.java

```diff
@@ -0,0 +1,126 @@
+    public void process(final LGraph graph, final IElkProgressMonitor progressMonitor) {
+        int layerIndex = 0;
+        for (Layer layer : graph) {
+            for (LNode node : layer.getNodes()) {
+                final int previousLayerIndex = layerIndex - 1;
+                Collections.sort(node.getPorts(), (p1, p2) -> {
+                    // Sort incoming edges by sorting their ports by the order of the nodes they connect to.
+                    if (p1.getOutgoingEdges().isEmpty() && p2.getOutgoingEdges().isEmpty()) {
+                        Layer previousLayer = graph.getLayers().get(previousLayerIndex);
+                        LNode p1Node = p1.getIncomingEdges().get(0).getSource().getNode();
+                        LNode p2Node = p2.getIncomingEdges().get(0).getSource().getNode();
+                        
+                        for (LNode previousNode : previousLayer) {
+                            if (previousNode.equals(p1Node)) {
+                                return 1;
+                            }
+                            if (previousNode.equals(p2Node)) {
+                                return -1;
+                            }
+                        }
+                    }
+
+                    // Sort outgoing edges by sorting their ports.
+                    if (p1.getIncomingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return Integer.compare(
+                                p1.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER),
+                                p2.getOutgoingEdges().get(0).getProperty(InternalProperties.MODEL_ORDER));
+                    }
+                    // Sort outgoing ports are ordered after incoming ports.
+                    if (p1.getOutgoingEdges().isEmpty() && p2.getIncomingEdges().isEmpty()) {
+                        return 1;
+                    } else {
+                        return -1;
+                    }
+                });
+            }
+            // Sort nodes.
+            Collections.sort(layer.getNodes(), (n1, n2) -> {
+                // If no model order is set, the one node is a dummy node and the nodes should be ordered
+                // by the connected edges.
+                if (!n1.hasProperty(InternalProperties.MODEL_ORDER)
+                        || !n2.hasProperty(InternalProperties.MODEL_ORDER)) {
+                    return Integer.compare(
+                            getModelOrderFromConnectedEdges(n1),
+                            getModelOrderFromConnectedEdges(n2));
+                }
+                // Order nodes by their order in the model.
+                return Integer.compare(
+                        n1.getProperty(InternalProperties.MODEL_ORDER),
+                        n2.getProperty(InternalProperties.MODEL_ORDER));
+            });
+            layerIndex++;
+        }
+    }
+    
+    /**
+     * The model order of the first incoming edge of a node.
```

Comment:
What happens if that doesn't have a model order set?

Reply:
All edges have a model order set, if they are processed by the `ElkGraphImporter`. Is it possible that such edges exist?